features.md

```diff
@@ -138,6 +138,18 @@ Annotations MUST be a key-value map that follows the same convention as the Key
 }
 ```
 
+## <a name="featuresUnsafeAnnotationPrefixes" />Unsafe annotation prefixes
+
+**`unsafeAnnotationPrefixes`** (array of strings, OPTIONAL) contains prefixes of the [`annotations` property of `config.json`](config.md#annotations)
+that may change the behavior of the runtime.
```

Comment:
I think a description of how the prefix should be interpreted needs to be put here. My personal opinion is we should have this be called `unsafeAnnotations` and have the ability to specify the full annotation name (`a.b.c` where only `a.b.c` is considered unsafe) as well as a namespace (`x.y.` where anything under `x.y` such as `x.y.a` and `x.y.b` but not `x.y` itself is considered unsafe).

Also this needs to be on one line to match the spec style.

Reply:
Also this definition is far too vague -- is the issue with these annotations that they could be used to bypass container protections (the original discussion we had about this issue on the security list), or that they are being parsed by the runtime at all?

The text implies the latter but that would mean:

1. Even something like `STOPSIGNAL` (`org.opencontainers.image.stopSignal`) would be considered "unsafe".
2. The purpose of using annotations to affect container configuration would be considered "unsafe" entirely, which might be true for some things (such as annotations that modify hook or systemd configurations) but is not true for others (a hypothetical generic NVIDIA GPU configuration, or `STOPSIGNAL`, for instance).

Maybe we should have two lists, one for annotations that the runtime parses at all (what the current text implies) and another for ones it considers unsafe (which we could define to mean that the runtime believes an attacker being able to control the value would be able to cause significant harm to the system). I think both lists would be useful for different reasons (the former is what annotation-based features are supported, the latter is which annotations are considered unsafe by the runtime).

Reply:
> My personal opinion is we should have this be called unsafeAnnotations and have the ability to specify the full annotation name (a.b.c where only a.b.c is considered unsafe) 

"a.b.c" is already valid input as an entry of `unsafeAnnotationPrefixes`.
A drawback is that you can't mark "a.b.c.d" safe, but probably it is not a huge deal in practice.

> `org.opencontainers.image.stopSignal`

This annotation is consumed by Engines, not by Runtimes.

> Maybe we should have two lists, one for annotations that the runtime parses at all (what the current text implies) and another for ones it considers unsafe (which we could define to mean that the runtime believes an attacker being able to control the value would be able to cause significant harm to the system). 

Seems a breaking change.

Reply:
> This annotation is consumed by Engines, not by Runtimes.

This is a distinction that doesn't exist in the runtime-spec (the word "engine" doesn't appear anywhere in the runtime-spec) -- if someone proposed that `runc kill` used the value of `org.opencontainers.image.stopSignal` by default, I wouldn't have an issue in principle with that idea (other than backwards-compatibility worries).

> Generating the precise list of the annotations might be sometimes difficult, e.g., for systemd annotations

If they can't be precise, a runtime can choose to be more conservative and output a blanket ban for all annotations related to systemd. The reason I wanted to have two lists is because it will let users choose how strict they want to be -- if they are running untrusted code in a very critical environment, they will probably want to reject any annotations that runc will parse _at all_ -- while most users will just trust that the runtime has a good idea of what annotations are unsafe.